File: src/scanner.test.ts
import { test } from "node:test";
import { expect } from "expect";

import { createScanner } from "./scanner.ts";
import { SyntaxKind, TokenFlags } from "./types.ts";

interface ScannedToken {
  kind: SyntaxKind;
  value: string;
  text: string;
  start: number;
  end: number;
  flags: TokenFlags;
  precedingLineBreak: boolean;
}

function tokenize(src: string): { tokens: ScannedToken[]; errorCodes: number[] } {
  const errorCodes: number[] = [];
  const scanner = createScanner(src, message => errorCodes.push(message.code));
  const tokens: ScannedToken[] = [];
  let kind = scanner.scan();
  while (kind !== SyntaxKind.EndOfFileToken) {
    tokens.push({
      kind,
      value: scanner.getTokenValue(),
      text: scanner.getTokenText(),
      start: scanner.getTokenStart(),
      end: scanner.getTokenEnd(),
      flags: scanner.getTokenFlags(),
      precedingLineBreak: scanner.hasPrecedingLineBreak(),
    });
    kind = scanner.scan();
  }
  return { tokens, errorCodes };
}

function kinds(src: string): SyntaxKind[] {
  return tokenize(src).tokens.map(t => t.kind);
}

test("keywords vs identifiers", () => {
  const { tokens } = tokenize("class Foo int var");
  expect(tokens.map(t => t.kind)).toEqual([
    SyntaxKind.ClassKeyword,
    SyntaxKind.Identifier,
    SyntaxKind.IntKeyword,
    // 'var' is contextual -> scanned as an identifier
    SyntaxKind.Identifier,
  ]);
  expect(tokens[1]!.value).toBe("Foo");
  expect(tokens[3]!.value).toBe("var");
});

test("identifiers allow $ and _", () => {
  expect(kinds("$x _y a1 _")).toEqual([
    SyntaxKind.Identifier,
    SyntaxKind.Identifier,
    SyntaxKind.Identifier,
    SyntaxKind.Identifier,
  ]);
});

test("operators use maximal munch", () => {
  expect(kinds("+ ++ += -> :: ... < <= << <<= == != >>>=")).toEqual([
    SyntaxKind.PlusToken,
    SyntaxKind.PlusPlusToken,
    SyntaxKind.PlusEqualsToken,
    SyntaxKind.ArrowToken,
    SyntaxKind.ColonColonToken,
    SyntaxKind.DotDotDotToken,
    SyntaxKind.LessThanToken,
    SyntaxKind.LessThanEqualsToken,
    SyntaxKind.LessThanLessThanToken,
    SyntaxKind.LessThanLessThanEqualsToken,
    SyntaxKind.EqualsEqualsToken,
    SyntaxKind.ExclamationEqualsToken,
    // '>' family stays single until reScanGreaterToken, so ">>>=" is four tokens
    SyntaxKind.GreaterThanToken,
    SyntaxKind.GreaterThanToken,
    SyntaxKind.GreaterThanToken,
    SyntaxKind.EqualsToken,
  ]);
});

test("'>' is scanned one at a time so nested generics close cleanly", () => {
  // List<List<T>> -> ... T > >  (two single GreaterThanToken)
  const ks = kinds("List<List<T>>");
  expect(ks.slice(-2)).toEqual([SyntaxKind.GreaterThanToken, SyntaxKind.GreaterThanToken]);
});

test("reScanGreaterToken merges the '>' family on demand", () => {
  const cases: Array<[string, SyntaxKind]> = [
    [">>", SyntaxKind.GreaterThanGreaterThanToken],
    [">>>", SyntaxKind.GreaterThanGreaterThanGreaterThanToken],
    [">=", SyntaxKind.GreaterThanEqualsToken],
    [">>=", SyntaxKind.GreaterThanGreaterThanEqualsToken],
    [">>>=", SyntaxKind.GreaterThanGreaterThanGreaterThanEqualsToken],
  ];
  for (const [src, expected] of cases) {
    const scanner = createScanner(src);
    expect(scanner.scan()).toBe(SyntaxKind.GreaterThanToken);
    expect(scanner.reScanGreaterToken()).toBe(expected);
    expect(scanner.getTokenEnd()).toBe(src.length);
  }
});

test("numeric literals and their flags", () => {
  const { tokens } = tokenize("0 42 0xFF 0b1010 0777 1_000 3.14 1.0f 2.0d 100L 1e10");
  expect(tokens.every(t => t.kind === SyntaxKind.NumericLiteral)).toBe(true);
  const flag = (i: number) => tokens[i]!.flags;
  expect(flag(2) & TokenFlags.HexSpecifier).toBeTruthy();
  expect(flag(3) & TokenFlags.BinarySpecifier).toBeTruthy();
  expect(flag(4) & TokenFlags.OctalSpecifier).toBeTruthy();
  expect(flag(5) & TokenFlags.ContainsUnderscore).toBeTruthy();
  expect(flag(7) & TokenFlags.FloatSuffix).toBeTruthy();
  expect(flag(8) & TokenFlags.DoubleSuffix).toBeTruthy();
  expect(flag(9) & TokenFlags.LongSuffix).toBeTruthy();
});

test("leading-dot float", () => {
  const { tokens } = tokenize(".5");
  expect(tokens).toHaveLength(1);
  expect(tokens[0]!.kind).toBe(SyntaxKind.NumericLiteral);
  expect(tokens[0]!.value).toBe(".5");
});

test("string literal decodes escapes into the token value", () => {
  const { tokens } = tokenize('"a\\tb\\n\\"c\\u0041"');
  expect(tokens[0]!.kind).toBe(SyntaxKind.StringLiteral);
  expect(tokens[0]!.value).toBe('a\tb\n"cA');
});

test("character literal", () => {
  const { tokens } = tokenize("'a' '\\n'");
  expect(tokens.map(t => t.kind)).toEqual([
    SyntaxKind.CharacterLiteral,
    SyntaxKind.CharacterLiteral,
  ]);
  expect(tokens[0]!.value).toBe("a");
  expect(tokens[1]!.value).toBe("\n");
});

test("text block", () => {
  const src = '"""\nhello\n"""';
  const { tokens, errorCodes } = tokenize(src);
  expect(errorCodes).toEqual([]);
  expect(tokens).toHaveLength(1);
  expect(tokens[0]!.kind).toBe(SyntaxKind.TextBlockLiteral);
});

test("trivia is skipped and line breaks are flagged", () => {
  const { tokens } = tokenize("a // comment\nb /* multi\nline */ c");
  expect(tokens.map(t => t.kind)).toEqual([
    SyntaxKind.Identifier,
    SyntaxKind.Identifier,
    SyntaxKind.Identifier,
  ]);
  expect(tokens[0]!.precedingLineBreak).toBe(false);
  expect(tokens[1]!.precedingLineBreak).toBe(true);
  expect(tokens[2]!.precedingLineBreak).toBe(true); // newline inside the block comment
});

test("token positions exclude leading trivia", () => {
  const { tokens } = tokenize("  foo");
  expect(tokens[0]!.start).toBe(2);
  expect(tokens[0]!.end).toBe(5);
});

test("unterminated string reports an error", () => {
  const { tokens, errorCodes } = tokenize('"abc');
  expect(tokens[0]!.kind).toBe(SyntaxKind.StringLiteral);
  expect(tokens[0]!.flags & TokenFlags.Unterminated).toBeTruthy();
  expect(errorCodes).toHaveLength(1);
});

test("unterminated block comment reports an error", () => {
  const { errorCodes } = tokenize("/* never closed");
  expect(errorCodes).toHaveLength(1);
});

test("invalid character reports an error and does not loop", () => {
  const { tokens, errorCodes } = tokenize("a # b");
  expect(errorCodes).toHaveLength(1);
  expect(tokens.map(t => t.kind)).toEqual([
    SyntaxKind.Identifier,
    SyntaxKind.Unknown,
    SyntaxKind.Identifier,
  ]);
});

test("lookAhead restores scanner state", () => {
  const scanner = createScanner("a b");
  expect(scanner.scan()).toBe(SyntaxKind.Identifier);
  const peeked = scanner.lookAhead(() => scanner.scan());
  expect(peeked).toBe(SyntaxKind.Identifier);
  // after lookAhead we are still positioned on 'a'
  expect(scanner.getTokenValue()).toBe("a");
  expect(scanner.scan()).toBe(SyntaxKind.Identifier);
  expect(scanner.getTokenValue()).toBe("b");
});

test("unicode escapes and underscores in literals", () => {
  const { tokens } = tokenize('"\\u0041\\u0042" 0xFF_FF 0b1010_1010 1_000_000L');
  expect(tokens[0]!.value).toBe("AB");
  expect(tokens[1]!.flags & TokenFlags.HexSpecifier).toBeTruthy();
  expect(tokens[1]!.flags & TokenFlags.ContainsUnderscore).toBeTruthy();
  expect(tokens[3]!.flags & TokenFlags.LongSuffix).toBeTruthy();
});

test("CRLF line breaks set the preceding-line-break flag", () => {
  const { tokens } = tokenize("a\r\nb");
  expect(tokens[1]!.precedingLineBreak).toBe(true);
});

test("character literal with octal and unicode escapes", () => {
  const { tokens } = tokenize("'\\101' '\\u0041'");
  expect(tokens[0]!.value).toBe("A"); // \101 octal = 65 = 'A'
  expect(tokens[1]!.value).toBe("A");
});

test("consecutive '<' are scanned as separate tokens", () => {
  expect(kinds("a < < b")).toEqual([
    SyntaxKind.Identifier,
    SyntaxKind.LessThanToken,
    SyntaxKind.LessThanToken,
    SyntaxKind.Identifier,
  ]);
});

test("hexadecimal floating-point literals (0x1p1023, 0x1.8p-3)", () => {
  const { tokens, errorCodes } = tokenize("0x1p1023 0x1.8p-3 0x1.0p0d");
  expect(errorCodes).toEqual([]);
  expect(tokens.map(t => t.kind)).toEqual([
    SyntaxKind.NumericLiteral,
    SyntaxKind.NumericLiteral,
    SyntaxKind.NumericLiteral,
  ]);
  expect(tokens[2]!.flags & TokenFlags.DoubleSuffix).toBeTruthy();
});

File: togo/internal/compiler/scanner_test.go
package compiler

import (
	"reflect"
	"testing"
)

type scannedToken struct {
	kind               SyntaxKind
	value              string
	text               string
	start, end         int
	flags              TokenFlags
	precedingLineBreak bool
}

func tokenize(src string) ([]scannedToken, []int) {
	var codes []int
	s := NewScanner(src, func(m DiagnosticMessage, pos, length int) { codes = append(codes, int(m.Code)) })
	var tokens []scannedToken
	for kind := s.Scan(); kind != EndOfFileToken; kind = s.Scan() {
		tokens = append(tokens, scannedToken{
			kind: kind, value: s.TokenValue(), text: s.TokenText(),
			start: s.TokenStart(), end: s.TokenEnd(), flags: s.TokenFlags(),
			precedingLineBreak: s.HasPrecedingLineBreak(),
		})
	}
	return tokens, codes
}

func kindsOf(src string) []SyntaxKind {
	tokens, _ := tokenize(src)
	ks := make([]SyntaxKind, len(tokens))
	for i, t := range tokens {
		ks[i] = t.kind
	}
	return ks
}

func eq(t *testing.T, label string, got, want any) {
	t.Helper()
	if !reflect.DeepEqual(got, want) {
		t.Errorf("%s: got %v, want %v", label, got, want)
	}
}

// --- port of scanner.test.ts -------------------------------------------------

func TestKeywordsVsIdentifiers(t *testing.T) {
	tokens, _ := tokenize("class Foo int var")
	eq(t, "kinds", []SyntaxKind{tokens[0].kind, tokens[1].kind, tokens[2].kind, tokens[3].kind},
		[]SyntaxKind{ClassKeyword, Identifier, IntKeyword, Identifier})
	eq(t, "Foo", tokens[1].value, "Foo")
	eq(t, "var", tokens[3].value, "var")
}

func TestIdentifiersDollarUnderscore(t *testing.T) {
	eq(t, "kinds", kindsOf("$x _y a1 _"), []SyntaxKind{Identifier, Identifier, Identifier, Identifier})
}

func TestOperatorsMaximalMunch(t *testing.T) {
	eq(t, "kinds", kindsOf("+ ++ += -> :: ... < <= << <<= == != >>>="), []SyntaxKind{
		PlusToken, PlusPlusToken, PlusEqualsToken, ArrowToken, ColonColonToken, DotDotDotToken,
		LessThanToken, LessThanEqualsToken, LessThanLessThanToken, LessThanLessThanEqualsToken,
		EqualsEqualsToken, ExclamationEqualsToken,
		GreaterThanToken, GreaterThanToken, GreaterThanToken, EqualsToken,
	})
}

func TestNestedGenericsCloseCleanly(t *testing.T) {
	ks := kindsOf("List<List<T>>")
	eq(t, "last two", ks[len(ks)-2:], []SyntaxKind{GreaterThanToken, GreaterThanToken})
}

func TestReScanGreaterToken(t *testing.T) {
	cases := []struct {
		src      string
		expected SyntaxKind
	}{
		{">>", GreaterThanGreaterThanToken},
		{">>>", GreaterThanGreaterThanGreaterThanToken},
		{">=", GreaterThanEqualsToken},
		{">>=", GreaterThanGreaterThanEqualsToken},
		{">>>=", GreaterThanGreaterThanGreaterThanEqualsToken},
	}
	for _, c := range cases {
		s := NewScanner(c.src, nil)
		if s.Scan() != GreaterThanToken {
			t.Fatalf("%q: first token not '>'", c.src)
		}
		if got := s.ReScanGreaterToken(); got != c.expected {
			t.Errorf("%q: reScan = %v, want %v", c.src, got, c.expected)
		}
		if s.TokenEnd() != len(c.src) {
			t.Errorf("%q: end = %d, want %d", c.src, s.TokenEnd(), len(c.src))
		}
	}
}

func TestNumericFlags(t *testing.T) {
	tokens, _ := tokenize("0 42 0xFF 0b1010 0777 1_000 3.14 1.0f 2.0d 100L 1e10")
	for _, tok := range tokens {
		if tok.kind != NumericLiteral {
			t.Fatalf("non-numeric token: %v", tok.kind)
		}
	}
	check := func(i int, flag TokenFlags) {
		if tokens[i].flags&flag == 0 {
			t.Errorf("token %d (%q) missing flag %d", i, tokens[i].text, flag)
		}
	}
	check(2, HexSpecifier)
	check(3, BinarySpecifier)
	check(4, OctalSpecifier)
	check(5, ContainsUnderscore)
	check(7, FloatSuffix)
	check(8, DoubleSuffix)
	check(9, LongSuffix)
}

func TestLeadingDotFloat(t *testing.T) {
	tokens, _ := tokenize(".5")
	if len(tokens) != 1 || tokens[0].kind != NumericLiteral || tokens[0].value != ".5" {
		t.Errorf("got %+v", tokens)
	}
}

func TestStringEscapes(t *testing.T) {
	tokens, _ := tokenize(`"a\tb\n\"c\u0041"`)
	eq(t, "kind", tokens[0].kind, StringLiteral)
	eq(t, "value", tokens[0].value, "a\tb\n\"cA")
}

func TestCharacterLiteral(t *testing.T) {
	tokens, _ := tokenize(`'a' '\n'`)
	eq(t, "kinds", []SyntaxKind{tokens[0].kind, tokens[1].kind}, []SyntaxKind{CharacterLiteral, CharacterLiteral})
	eq(t, "a", tokens[0].value, "a")
	eq(t, "newline", tokens[1].value, "\n")
}

func TestTextBlock(t *testing.T) {
	tokens, codes := tokenize("\"\"\"\nhello\n\"\"\"")
	eq(t, "codes", codes, []int(nil))
	if len(tokens) != 1 || tokens[0].kind != TextBlockLiteral {
		t.Errorf("got %+v", tokens)
	}
}

func TestTriviaAndLineBreaks(t *testing.T) {
	tokens, _ := tokenize("a // comment\nb /* multi\nline */ c")
	eq(t, "kinds", []SyntaxKind{tokens[0].kind, tokens[1].kind, tokens[2].kind}, []SyntaxKind{Identifier, Identifier, Identifier})
	eq(t, "t0 lb", tokens[0].precedingLineBreak, false)
	eq(t, "t1 lb", tokens[1].precedingLineBreak, true)
	eq(t, "t2 lb", tokens[2].precedingLineBreak, true)
}

func TestTokenPositionsExcludeTrivia(t *testing.T) {
	tokens, _ := tokenize("  foo")
	eq(t, "start", tokens[0].start, 2)
	eq(t, "end", tokens[0].end, 5)
}

func TestUnterminatedString(t *testing.T) {
	tokens, codes := tokenize(`"abc`)
	eq(t, "kind", tokens[0].kind, StringLiteral)
	if tokens[0].flags&Unterminated == 0 {
		t.Error("expected Unterminated flag")
	}
	eq(t, "codes len", len(codes), 1)
}

func TestUnterminatedBlockComment(t *testing.T) {
	_, codes := tokenize("/* never closed")
	eq(t, "codes len", len(codes), 1)
}

func TestInvalidCharacter(t *testing.T) {
	tokens, codes := tokenize("a # b")
	eq(t, "codes len", len(codes), 1)
	eq(t, "kinds", []SyntaxKind{tokens[0].kind, tokens[1].kind, tokens[2].kind}, []SyntaxKind{Identifier, Unknown, Identifier})
}

func TestLookAheadRestores(t *testing.T) {
	s := NewScanner("a b", nil)
	eq(t, "first", s.Scan(), Identifier)
	peeked := LookAhead(s, func() SyntaxKind { return s.Scan() })
	eq(t, "peeked", peeked, Identifier)
	eq(t, "still a", s.TokenValue(), "a")
	eq(t, "next", s.Scan(), Identifier)
	eq(t, "now b", s.TokenValue(), "b")
}

func TestUnicodeEscapesAndUnderscores(t *testing.T) {
	tokens, _ := tokenize(`"\u0041\u0042" 0xFF_FF 0b1010_1010 1_000_000L`)
	eq(t, "AB", tokens[0].value, "AB")
	if tokens[1].flags&HexSpecifier == 0 || tokens[1].flags&ContainsUnderscore == 0 {
		t.Error("hex/underscore flags missing")
	}
	if tokens[3].flags&LongSuffix == 0 {
		t.Error("long suffix missing")
	}
}

func TestCRLFLineBreak(t *testing.T) {
	tokens, _ := tokenize("a\r\nb")
	eq(t, "t1 lb", tokens[1].precedingLineBreak, true)
}

func TestCharOctalUnicodeEscapes(t *testing.T) {
	tokens, _ := tokenize(`'\101' '\u0041'`)
	eq(t, "octal A", tokens[0].value, "A")
	eq(t, "unicode A", tokens[1].value, "A")
}

func TestConsecutiveLessThan(t *testing.T) {
	eq(t, "kinds", kindsOf("a < < b"), []SyntaxKind{Identifier, LessThanToken, LessThanToken, Identifier})
}

func TestStringEscapeEdgeCases(t *testing.T) {
	// \s is a space (SE15); multiple u's may precede the 4 hex digits; an
	// unknown escape passes the char through; \' decodes to a single quote.
	tokens, codes := tokenize(`"a\sb\uuuu0041\q\'"`)
	eq(t, "codes", codes, []int(nil))
	eq(t, "value", tokens[0].value, "a bAq'")
}

func TestStringEscapeOctalCutoff(t *testing.T) {
	// \777 exceeds 0xff, so only \77 (= 0x3f = '?') is consumed and the
	// trailing 7 is a literal character.
	tokens, _ := tokenize(`"\777"`)
	eq(t, "value", tokens[0].value, "?7")
}

func TestIncompleteUnicodeEscape(t *testing.T) {
	// Fewer than 4 hex digits after \u reports a diagnostic.
	tokens, codes := tokenize(`"\u12"`)
	eq(t, "kind", tokens[0].kind, StringLiteral)
	eq(t, "codes", codes, []int{1104}) // HexadecimalDigitExpected
}

func TestTrailingBackslashAtEOF(t *testing.T) {
	// A backslash with nothing after it decodes to a literal backslash; the
	// string is also unterminated.
	tokens, _ := tokenize("\"\\")
	eq(t, "value", tokens[0].value, "\\")
	if tokens[0].flags&Unterminated == 0 {
		t.Error("expected Unterminated flag")
	}
}

func TestHexFloats(t *testing.T) {
	tokens, codes := tokenize("0x1p1023 0x1.8p-3 0x1.0p0d")
	eq(t, "codes", codes, []int(nil))
	eq(t, "kinds", []SyntaxKind{tokens[0].kind, tokens[1].kind, tokens[2].kind}, []SyntaxKind{NumericLiteral, NumericLiteral, NumericLiteral})
	if tokens[2].flags&DoubleSuffix == 0 {
		t.Error("double suffix missing")
	}
}

// --- port of scanner.edge.test.ts --------------------------------------------

func TestHexFloatNegativeExponent(t *testing.T) {
	tokens, codes := tokenize("0x1p-3 0x.8p1 0x1.8p3f")
	eq(t, "codes", codes, []int(nil))
	for _, tok := range tokens {
		if tok.kind != NumericLiteral {
			t.Fatalf("non-numeric: %v", tok.kind)
		}
	}
	if tokens[2].flags&FloatSuffix == 0 {
		t.Error("float suffix missing")
	}
}

func TestUnderscoresInFractionExponent(t *testing.T) {
	tokens, codes := tokenize("1_000.000_1 1_0e1_0")
	eq(t, "codes", codes, []int(nil))
	eq(t, "len", len(tokens), 2)
	if tokens[0].flags&ContainsUnderscore == 0 {
		t.Error("underscore flag missing")
	}
}

func TestLeadingZeroOctal(t *testing.T) {
	tokens, _ := tokenize("0 00 0777 0L")
	if tokens[0].flags&OctalSpecifier != 0 {
		t.Error("plain 0 should not be octal")
	}
	if tokens[1].flags&OctalSpecifier == 0 || tokens[2].flags&OctalSpecifier == 0 {
		t.Error("00 / 0777 should be octal")
	}
	if tokens[3].flags&LongSuffix == 0 {
		t.Error("0L should be long")
	}
}

func TestCharEscapesDecode(t *testing.T) {
	tokens, _ := tokenize(`'A' '\377' '\0'`)
	eq(t, "u0041", tokens[0].value, "A")
	eq(t, "377", tokens[1].value, string(rune(255)))
	eq(t, "0", tokens[2].value, string(rune(0))) // \0 octal = NUL

}

func TestTextBlockEmbeddedQuotes(t *testing.T) {
	tokens, codes := tokenize("\"\"\"\nsay \"hi\" there\n\"\"\"")
	eq(t, "codes", codes, []int(nil))
	if len(tokens) != 1 || tokens[0].kind != TextBlockLiteral {
		t.Errorf("got %+v", tokens)
	}
}

func TestBlockCommentNoNesting(t *testing.T) {
	eq(t, "kinds", kindsOf("/* a /* still */ x */ y"), []SyntaxKind{Identifier, AsteriskToken, SlashToken, Identifier})
}

func TestCompoundShiftAssign(t *testing.T) {
	eq(t, "kinds", kindsOf("<<="), []SyntaxKind{LessThanLessThanEqualsToken})
}

func TestUnderscoreDollarIdentifiers(t *testing.T) {
	want := []SyntaxKind{Identifier, Identifier, Identifier, Identifier, Identifier}
	eq(t, "kinds", kindsOf("_ $ $a a$b _x_"), want)
}

func TestArrowColonVarargsAt(t *testing.T) {
	eq(t, "kinds", kindsOf("-> :: ... @"), []SyntaxKind{ArrowToken, ColonColonToken, DotDotDotToken, AtToken})
}

func TestStringSimpleEscapes(t *testing.T) {
	tokens, _ := tokenize(`"\b\t\n\f\r\"\\"`)
	eq(t, "value", tokens[0].value, "\b\t\n\f\r\"\\")
}

func TestEmptyAndWhitespace(t *testing.T) {
	eq(t, "empty", kindsOf(""), []SyntaxKind{})
	eq(t, "whitespace", kindsOf("   \n\t  "), []SyntaxKind{})
}

func TestErrorCodes(t *testing.T) {
	cases := []struct {
		src  string
		want []int
	}{
		{"'a", []int{1101}}, {"'", []int{1101}},
		{"0x", []int{1104}}, {"0xG", []int{1104}},
		{"0b", []int{1105}}, {"0b2", []int{1105}},
		{"1e", []int{1103}}, {"0x1p", []int{1103}},
	}
	for _, c := range cases {
		_, codes := tokenize(c.src)
		eq(t, c.src, codes, c.want)
	}
}
